sopel/tools/identifiers.py

```diff
@@ -0,0 +1,171 @@
+RFC1459_TABLE = str.maketrans(
+    string.ascii_uppercase + '[]\\~',
+    string.ascii_lowercase + '{}|^',
+)
+RFC1459_STRICT_TABLE = str.maketrans(
+    string.ascii_uppercase + '[]\\',
+    string.ascii_lowercase + '{}|',
+)
+
+
+def ascii_lower(text: str) -> str:
+    """Lower ```text``` according to the ASCII CASEMAPPING"""
+    return text.translate(ASCII_TABLE)
+
+
+def rfc1459_lower(text: str) -> str:
+    """Lower ``text`` according to :rfc:`1459` (with ``~`` mapped to ``^``).
+
+    Similar function as :func:`rfc1459_strict_lower` that also maps ``~`` to
+    ``^`` as defined for the ``rfc1459`` value of the
+    `CASEMAPPING parameter`__.
+
+    .. __: https://modern.ircdocs.horse/index.html#casemapping-parameter
+    """
+    return text.translate(RFC1459_TABLE)
+
+
+def rfc1459_strict_lower(text: str) -> str:
+    """Lower ``text`` according to :rfc:`1459` (strict version).
+
+    As per `section 2.2`__:
+
+        Because of IRC's scandanavian origin, the characters ``{}|`` are
+        considered to be the lower case equivalents of the characters ``[]\\``.
+
+    .. __: https://datatracker.ietf.org/doc/html/rfc1459#section-2.2
+    """
+    return text.translate(RFC1459_STRICT_TABLE)
+
+
+_channel_prefixes = ('#', '&', '+', '!')
+
+
+class Identifier(str):
+    """A ``str`` subclass which acts appropriately for IRC identifiers.
+
+    When used as normal ``str`` objects, case will be preserved.
+    However, when comparing two Identifier objects, or comparing a Identifier
+    object with a ``str`` object, the comparison will be case insensitive.
+    This case insensitivity includes the case convention conventions regarding
+    ``[]``, ``{}``, ``|``, ``\\``, ``^`` and ``~`` described in RFC 2812.
+    """
+    def __new__(cls, *args, **kwargs) -> 'Identifier':
+        return str.__new__(cls, *args)
+
+    def __init__(
+        self,
+        identifier: str,
+        *,
+        casemapping: Casemapping = rfc1459_lower,
+    ) -> None:
+        super().__init__()
+        self.casemapping: Casemapping = casemapping
+        """Casemapping function to lower the identifier."""
+        self._lowered = self.casemapping(identifier)
+
+    def lower(self):
+        """Get the RFC 2812-compliant lowercase version of this identifier.
+
+        :return: RFC 2812-compliant lowercase version of the
+                 :py:class:`Identifier` instance
+        :rtype: str
+        """
+        return self.casemapping(self)
+
+    @staticmethod
+    def _lower(identifier: str):
+        """Convert an identifier to lowercase per RFC 2812.
+
+        :param str identifier: the identifier (nickname or channel) to convert
+        :return: RFC 2812-compliant lowercase version of ``identifier``
+        :rtype: str
+        """
+        if isinstance(identifier, Identifier):
+            return identifier.lower()
+        return rfc1459_lower(identifier)
+
+    @staticmethod
+    def _lower_swapped(identifier: str):
```

Comment:
Oh, yeah, right. I know why I didn't do it yet: because I want to move the `deprecated` function somewhere else, if you agree: I think `sopel.deprecated` or `sopel.utils` would be good place for that one. I actually think that `from sopel.deprecated import deprecated` works really nicely!